src/protobuf/buffer.cr

```diff
@@ -210,7 +241,13 @@ module Protobuf
     def write_packed(arr, pb_type)
       io = IO::Memory.new
       tmp_buf = self.class.new(io)
-      arr.not_nil!.each {|i| tmp_buf.write(i, pb_type) }
+      arr.not_nil!.each do |i|
+        if i.responds_to? :to_protobuf
+          i.to_protobuf(io)
+        else
+          tmp_buf.write(i, pb_type)
+        end
```

Comment:
I like the duck-typing happening here, but I wonder if we could get the Crystal type system involved here because duck typing in Crystal is typically a bit more subtle.

Do you have any thoughts on something like this? I'm completely spitballing it so I'm not sure it'll work as written:

```crystal
def write(value, io : IO, pb_type)
  self.class.new(io).write(value, pb_type)
end

def write(message : Protobuf::Message, io : IO, pb_type)
  message.to_protobuf io
end
```

And then within the array, we just call invoke the overloaded method?